ha_gateway.rb

```diff
@@ -27,26 +23,205 @@
+        api.update_ufo(r,g,b,params['UFOWWLevel'].to_i*2.55, false)
+      
+      end
+    else # Adjust a bulb
+      # Adjust HSL
+      if params.include?('hue') && params.include?('saturation') && params.include?('level')
+        # If H & S from ST are the WW button's level, enable WW
+        if params['hue'] == "16.66666666666666" && params['saturation'] == "27.45097875595093"
+          w = params['level'].to_i*2.55 # Convert from 0-100 to 0-255
+          api.update_bulb_white(w,false)
+        else
+          r,g,b = hsvToRgb(params['hue'].to_f*3.6, params['saturation'].to_f, params['level'].to_f)
+          api.update_bulb_color(r,g,b, false)
+        end
+      end
+      # Adjust level only
+      if params.include?('level') && !params.include?('hue') && !params.include?('saturation') 
+        # If colors are off, adjust WW
+        if [r,g,b] == [0,0,0]
+          w = params['level'].to_i*2.55 #Convert from 0-100 to 0-255
+          api.update_bulb_white(w,false)
+        else
+          h,s,v = rgbToHsv(r,g,b)
+          r,g,b = hsvToRgb(h,s,params['level'].to_i*2.55)
+          api.update_bulb_color(r,g,b, false)
+        end
+      end
+      # Adjust saturation only
+      if params.include?('saturation') && !params.include?('hue') && !params.include?('level') 
+        # Get the bulb's hue and level, if in WW mode
+        if [r,g,b] == [0,0,0] && w > 0
+          r,g,b = hsvToRgb(60.0,27.45097875595093,w/2.55)
+        end
+        h,s,v = rgbToHsv(r,g,b)
+        r,g,b = hsvToRgb(h,params['saturation'].to_f,v)
+        api.update_bulb_color(r,g,b, false)
+      end
+      # Adjust hue only
+      if params.include?('hue') && !params.include?('saturation') && !params.include?('level') 
+        # Get the bulb's hue and level, if in WW mode
+        if [r,g,b] == [0,0,0] && w > 0
+          r,g,b = hsvToRgb(60.0,27.45097875595093,w/2.55)
+        end
+        h,s,v = rgbToHsv(r,g,b)
+        r,g,b = hsvToRgb(params['hue'].to_f,s,v)
+        api.update_bulb_color(r,g,b, false)
+      end
+    end
+
+    r,g,b,w,powerState = api.current_status # Get new settings from the device
+    h,s,l = rgbToHsv(r,g,b)
+
+    if !params['deviceType'].casecmp("UFO").zero? # If the device isn't a UFO, set the RGB to what WW's H & S are
+      if [r,g,b] == [0,0,0] && w > 0
+        r,g,b = hsvToRgb(16.66666666666666, 27.45097875595093, (w/2.55))
+      end
+      status 200
+      headers \
+        "powerState" => powerState,
+        "level" => l.to_s,
+        "hex" => "#" + to_hex(r) + to_hex(g) + to_hex(b)
+
+    else # If the device is a UFO, change the other parameters.
+      status 200
+      headers \
+        "powerState" => powerState,
+        "level" => l.to_s,
+        "hex" =>  "#" + to_hex(r) + to_hex(g) + to_hex(b),
+        "UFOWWLevel" => (w/2.55).to_s
+    end
+    status 200
+    body '{"success": true}'
+end
+def rgbToHsv(r, g, b)
```

Comment:
I'd be pretty surprised if it didn't support all of those. I found that HSV values were [0, 1] instead of [0, 255].